File: frontend/src/types/strategy.ts
export type StrategyExecutionMode = 'signal_only' | 'auto_trade'
export type StrategyStatus = 'draft' | 'active' | 'paused'
export type StrategySignalAction = 'buy' | 'sell' | 'reduce' | 'hold'
export type StrategySignalStrength = 'strong' | 'normal' | 'weak'
export type StrategyTargetType = 'single_symbol' | 'special_attention'

export interface StrategySignalPayload {
  signal: StrategySignalAction
  strength?: StrategySignalStrength | null
  trigger_reason?: string | null
  entry_price_ref?: number | null
  stop_loss_price?: number | null
  take_profit_price?: number | null
  position_pct?: number | null
  market_regime?: string | null
  requires_recommendation_confirmation?: boolean
  recommendation_confirmed?: boolean | null
  confirmation_source?: 'smart_selection' | 'special_attention_watchlist' | 'simulation_bypass' | 'none' | null
  recommendation_snapshot_date?: string | null
  position_add_path?: 'new_position' | 'first_add' | 'blocked_repeat_add' | null
  recommendation_score?: number | null
  recommendation_timing?: string | null
  execution_blockers?: string[]
  intraday_timing_status?: 'confirmed' | 'blocked' | 'unavailable' | 'disabled' | null
  intraday_trigger_reason?: string | null
  intraday_vwap?: number | null
  intraday_volume_ratio?: number | null
  intraday_latest_close?: number | null
  filter_passed?: boolean
  filter_reasons?: string[]
  trend_ok?: boolean | null
  volume_ok?: boolean | null
  volatility_ok?: boolean | null
  stretch_ok?: boolean | null
  market_regime_bias?: string | null
  strategy_metadata?: Record<string, unknown>
  strategy_context?: Record<string, unknown>
  standard_signal?: Record<string, unknown>
  [key: string]: string | number | boolean | string[] | Record<string, unknown> | null | undefined
}

export interface StrategyItem {
  id: number
  tenant_id: string
  name: string
  symbol: string
  target_type: StrategyTargetType
  target_config: Record<string, string | number | boolean>
  strategy_type: string
  status: StrategyStatus
  execution_mode: StrategyExecutionMode
  parameters: Record<string, number | string | boolean>
  latest_signal: StrategySignalAction
  latest_signal_summary: string | null
  readiness_status: 'draft' | 'observing' | 'paper_verified' | 'paused'
  readiness_summary: string | null
  signal_symbol: string
  signal_symbol_display: string | null
  resolved_target_count: number
  latest_run_status: string | null
  latest_run_at: string | null
  run_count_today: number
  total_run_count: number
  strategy_metadata: Record<string, unknown>
}

export interface StrategyVersionItem {
  id: number
  strategy_id: number
  version: number
  name: string
  symbol: string
  strategy_type: string
  execution_mode: StrategyExecutionMode
  target_type: StrategyTargetType
  target_config: Record<string, string | number | boolean>
  parameters: Record<string, unknown>
  created_at: string
}

export interface StrategyTemplateItem {
  key: string
  name: string
  category: string
  description: string
  scenario: string
  fit_for: string[]
  not_fit_for: string[]
  risk_note: string
  minimum_history: number
  auto_trade_allowed: boolean
  research_only: boolean
  parameter_bounds: Record<string, Record<string, unknown>>
  payload: {
    name: string
    symbol?: string | null
    target_type?: StrategyTargetType | null
    target_config?: Record<string, unknown> | null
    strategy_type: string
    execution_mode: StrategyExecutionMode
    parameters: Record<string, unknown>
  }
}

export interface StrategyCompareResult {
  items: Array<{
    key: string
    strategy_id: number
    version_id: number | null
    version: number | null
    name: string
    symbol: string
    strategy_type: string
    execution_mode: StrategyExecutionMode
    target_type: StrategyTargetType
    target_config: Record<string, unknown>
    parameters: Record<string, unknown>
    latest_run: Record<string, unknown> | null
    backtest_summary: Record<string, unknown>
    created_at: string | null
  }>
  parameter_diffs: Array<{ key: string; values: Record<string, unknown> }>
}

export interface StrategyRunItemResult {
  id: number
  symbol: string
  name: string | null
  signal: StrategySignalPayload
  order_submitted: boolean
  order_id: number | null
  order_status: string | null
  side: 'buy' | 'sell' | null
  quantity: number | null
  price: number | null
  reason: string | null
  strength: StrategySignalStrength | null
  trigger_reason: string | null
  stop_loss_price: number | null
  take_profit_price: number | null
  position_pct: number | null
  recommendation_confirmed: boolean | null
  confirmation_source: 'smart_selection' | 'special_attention_watchlist' | 'simulation_bypass' | 'none' | null
  recommendation_snapshot_date: string | null
  position_add_path: 'new_position' | 'first_add' | 'blocked_repeat_add' | null
  execution_blockers: string[]
  execution_environment: 'paper'
  created_at: string
}

export interface StrategyRunResult {
  id: number
  strategy_id: number
  status: string
  signal: StrategySignalPayload
  execution_mode: StrategyExecutionMode | null
  order_submitted: boolean
  order_id: number | null
  order_status: string | null
  side: 'buy' | 'sell' | null
  quantity: number | null
  price: number | null
  reason: string | null
  strength: StrategySignalStrength | null
  trigger_reason: string | null
  stop_loss_price: number | null
  take_profit_price: number | null
  position_pct: number | null
  recommendation_confirmed: boolean | null
  confirmation_source: 'smart_selection' | 'special_attention_watchlist' | 'simulation_bypass' | 'none' | null
  recommendation_snapshot_date: string | null
  position_add_path: 'new_position' | 'first_add' | 'blocked_repeat_add' | null
  execution_blockers: string[]
  execution_environment: 'paper'
  items: StrategyRunItemResult[]
  created_at: string
}

export interface StrategyRunHistory {
  runs: StrategyRunResult[]
}

File: backend/app/schemas/strategy.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class StrategyCreate(BaseModel):
    name: str
    symbol: str | None = None
    target_type: str | None = None
    target_config: dict[str, Any] | None = None
    strategy_type: str
    execution_mode: str = "signal_only"
    parameters: dict[str, Any] = Field(default_factory=dict)


class StrategyUpdate(BaseModel):
    name: str | None = None
    symbol: str | None = None
    target_type: str | None = None
    target_config: dict[str, Any] | None = None
    strategy_type: str | None = None
    status: str | None = None
    execution_mode: str | None = None
    parameters: dict[str, Any] | None = None


class StrategyRunItemRead(BaseModel):
    id: int
    symbol: str
    name: str | None = None
    signal: dict[str, Any]
    order_submitted: bool = False
    order_id: int | None = None
    order_status: str | None = None
    side: str | None = None
    quantity: int | None = None
    price: float | None = None
    reason: str | None = None
    strength: str | None = None
    trigger_reason: str | None = None
    stop_loss_price: float | None = None
    take_profit_price: float | None = None
    position_pct: float | None = None
    recommendation_confirmed: bool | None = None
    confirmation_source: str | None = None
    recommendation_snapshot_date: str | None = None
    position_add_path: str | None = None
    execution_blockers: list[str] = Field(default_factory=list)
    execution_environment: str = "paper"
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class StrategyRunRead(BaseModel):
    id: int
    strategy_id: int
    status: str
    signal: dict[str, Any]
    execution_mode: str | None = None
    order_submitted: bool = False
    order_id: int | None = None
    order_status: str | None = None
    side: str | None = None
    quantity: int | None = None
    price: float | None = None
    reason: str | None = None
    strength: str | None = None
    trigger_reason: str | None = None
    stop_loss_price: float | None = None
    take_profit_price: float | None = None
    position_pct: float | None = None
    recommendation_confirmed: bool | None = None
    confirmation_source: str | None = None
    recommendation_snapshot_date: str | None = None
    position_add_path: str | None = None
    execution_blockers: list[str] = Field(default_factory=list)
    execution_environment: str = "paper"
    items: list[StrategyRunItemRead] = Field(default_factory=list)
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class StrategyRunHistoryRead(BaseModel):
    runs: list[StrategyRunRead] = Field(default_factory=list)


class StrategyVersionRead(BaseModel):
    id: int
    strategy_id: int
    version: int
    name: str
    symbol: str
    strategy_type: str
    execution_mode: str
    target_type: str
    target_config: dict[str, Any]
    parameters: dict[str, Any]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class StrategyCompareRef(BaseModel):
    strategy_id: int
    version_id: int | None = None


class StrategyCompareRequest(BaseModel):
    items: list[StrategyCompareRef] = Field(min_length=2, max_length=4)


class StrategyCompareItemRead(BaseModel):
    key: str
    strategy_id: int
    version_id: int | None = None
    version: int | None = None
    name: str
    symbol: str
    strategy_type: str
    execution_mode: str
    target_type: str
    target_config: dict[str, Any]
    parameters: dict[str, Any]
    latest_run: dict[str, Any] | None = None
    backtest_summary: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime | None = None


class StrategyParameterDiffRead(BaseModel):
    key: str
    values: dict[str, Any]


class StrategyCompareRead(BaseModel):
    items: list[StrategyCompareItemRead]
    parameter_diffs: list[StrategyParameterDiffRead]


class StrategyDeleteRead(BaseModel):
    status: str
    id: int


class StrategyTemplateRead(BaseModel):
    key: str
    name: str
    category: str
    description: str
    scenario: str
    fit_for: list[str] = Field(default_factory=list)
    not_fit_for: list[str] = Field(default_factory=list)
    risk_note: str
    minimum_history: int
    auto_trade_allowed: bool
    research_only: bool = False
    parameter_bounds: dict[str, dict[str, Any]] = Field(default_factory=dict)
    payload: StrategyCreate


class StrategyRead(BaseModel):
    id: int
    tenant_id: str
    name: str
    symbol: str
    target_type: str
    target_config: dict[str, Any]
    strategy_type: str
    status: str
    execution_mode: str
    parameters: dict[str, Any]
    latest_signal: str
    latest_signal_summary: str | None = None
    readiness_status: str = "draft"
    readiness_summary: str | None = None
    signal_symbol: str
    signal_symbol_display: str | None = None
    resolved_target_count: int = 0
    latest_run_status: str | None = None
    latest_run_at: datetime | None = None
    run_count_today: int = 0
    total_run_count: int = 0
    strategy_metadata: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(from_attributes=True)
